subprojects/core/src/integTest/groovy/org/gradle/NameValidationIntegrationTest.groovy

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.executer.GradleContextualExecuter
+import org.gradle.integtests.fixtures.executer.GradleExecuter
+import org.gradle.test.fixtures.file.TestDirectoryProvider
+import org.gradle.test.fixtures.file.TestFile
+
+import static org.gradle.internal.os.OperatingSystem.WINDOWS
+import static org.gradle.internal.os.OperatingSystem.current
+
+class NameValidationIntegrationTest extends AbstractIntegrationSpec {
+
+    def setup() {
+        executer.expectDeprecationWarning()
+    }
+
+    def "project names should not contain forbidden characters"() {
+        when:
+        settingsFile << "rootProject.name = 'this::is::a::namespace'"
+        buildFile << ""
+
+        then:
+        succeeds 'help'
+    }
+
+    def "subproject names should not contain forbidden characters"() {
+        when:
+        settingsFile << "include 'folder:name with spaces'"
+
+        then:
+        succeeds 'help'
+        output.contains("The name 'name with spaces' contains at least one of the following characters:")
+    }
+
+    def "task names should not contain forbidden characters"() {
+        when:
+        buildFile << "task 'this/is/a/hierarchy'"
+
+        then:
+        succeeds 'this/is/a/hierarchy'
+    }
+
+    def "configuration names should not contain forbidden characters"() {
+        when:
+        buildFile << "configurations { 'some/really.\\\\strange name:' {} }"
+
+        then:
+        succeeds 'help'
+    }
+
+    def "does not assign an invalid project name from folder name"() {
+        given:
+        def buildFolder = file(current() == WINDOWS ? "folder  name" : "folder: name")
+        GradleExecuter subDirExecuter = new GradleContextualExecuter(distribution, new TestDirectoryProvider() {
```

Comment:
Cool. I did not find the `inDirectory()` method. Probably because I was expecting it to be called `with...`. Or I am just blind... 
Thanks for the hint! This makes the test code much simpler.